server/core/src/main/java/dev/slimevr/tracking/processor/Constraint.kt

```diff
@@ -0,0 +1,190 @@
+import kotlin.math.*
+
+/**
+ * Represents a function that applies a rotational constraint.
+ */
+typealias ConstraintFunction = (localRotation: Quaternion, limit1: Float, limit2: Float, limit3: Float) -> Quaternion
+
+/**
+ * Represents the rotational limits of a Bone relative to its parent,
+ * twist and swing are the max and min when constraintType is a hinge.
+ * Twist, swing, allowedDeviation, and maxDeviationFromTracker represent
+ * an angle in degrees.
+ */
+class Constraint(
+	val constraintType: ConstraintType,
+	twist: Float = 0.0f,
+	swing: Float = 0.0f,
+	allowedDeviation: Float = 0f,
+	maxDeviationFromTracker: Float = 15f,
+) {
+	private val constraintFunction = constraintTypeToFunc(constraintType)
+	private val twistRad = twist * FastMath.DEG_TO_RAD
+	private val swingRad = swing * FastMath.DEG_TO_RAD
+	private val allowedDeviationRad = allowedDeviation * FastMath.DEG_TO_RAD
+	private val maxDeviationFromTrackerRad = maxDeviationFromTracker * FastMath.DEG_TO_RAD
+
+	/**
+	 * allowModification may be false for reasons other than a tracker being on this bone
+	 * while hasTrackerRotation is only true if this bone has a tracker. These values are
+	 * to be used with an IK solver and are not currently set accurately
+	 */
+	var allowModifications = true
+	var hasTrackerRotation = false
+
+	/**
+	 * The rotation before any IK solve takes place. Again this value is not currently set accurately
+	 */
+	var initialRotation = Quaternion.IDENTITY
+
+	/**
+	 * Apply rotational constraints and if applicable force the rotation
+	 * to be unchanged unless it violates the constraints
+	 */
+	fun applyConstraint(rotation: Quaternion, thisBone: Bone): Quaternion {
+		// When constraints are being used during a IK solve the input rotation is not necessarily
+		// the bones global rotation, thus complete constraints must be specifically handled.
+		if (constraintType == ConstraintType.COMPLETE) return thisBone.getGlobalRotation()
+
+		// If there is no parent and this is not a complete constraint accept the rotation as is.
+		if (thisBone.parent == null) return rotation
+
+		val localRotation = getLocalRotation(rotation, thisBone)
+		val constrainedRotation = constraintFunction(localRotation, swingRad, twistRad, allowedDeviationRad)
+		return getWorldRotationFromLocal(constrainedRotation, thisBone)
+	}
+
+	/**
+	 * Force the given rotation to be within allowedDeviation degrees away from
+	 * initialRotation on both the twist and swing axis
+	 */
+	fun constrainToInitialRotation(rotation: Quaternion): Quaternion {
+		val rotationLocal = rotation * initialRotation.inv()
+		var (swingQ, twistQ) = decompose(rotationLocal, Vector3.NEG_Y)
+		swingQ = constrain(swingQ, maxDeviationFromTrackerRad)
+		twistQ = constrain(twistQ, maxDeviationFromTrackerRad)
+		return initialRotation * (swingQ * twistQ)
+	}
+
+	companion object {
+		const val ANGLE_THRESHOLD = 0.004f // == 0.25 degrees
+		const val FILTER_IMPACT_THRESHOLD = 0.0349 // == 2 degrees
```

Comment:
Was this supposed to be a double? 👀
```suggestion
		const val FILTER_IMPACT_THRESHOLD = 0.0349f // == 2 degrees
```